Failure pushing multi-arch images to Private AWS ECR with tag immutability enabled

Hi,

As per the instructions in [TROUBLESHOOTING.md for the docker/build-push-action GitHub Action](https://github.com/docker/build-push-action/blob/master/TROUBLESHOOTING.md), I am opening an issue here. If this is not the right place, please let me know. This could be an issue with the [ECR implementation of manifest lists.
](https://github.com/aws/containers-roadmap/issues/505)

**TL;DR**, from [Twitter](https://twitter.com/iamvlaaaaaaad/status/1365715640447889409):
> Pushing a multi-arch (AMD, ARM) container image to ECR:
>
> ECR Public: everything works 🎉 
> ECR Private, no tag immutability: everything works ✨ 
>
> ECR Private, tag immutability enabled: docker/build-push-action fails, containerd works 🙃 

--- 

Pushing a multi-arch (AMD, ARM) image to a private AWS ECR repository fails with `unexpected status: 400 Bad Request`. Pushing the same image to ECR, using containerd works.

I created [an example repository](https://github.com/Vlaaaaaaad/ecr-dual-arch-immutable-buildkit-bug) to show the issue and the relevant config/logs. You can see [the GitHub Action failing here](https://github.com/Vlaaaaaaad/ecr-dual-arch-immutable-buildkit-bug/runs/2004038467?check_suite_focus=true) and you can see [containerd working here](https://github.com/Vlaaaaaaad/ecr-dual-arch-immutable-buildkit-bug/runs/2004038465?check_suite_focus=true).

Relevant logs for the failure:
```
2021-03-01T13:06:04.8482250Z #21 exporting to image
2021-03-01T13:06:04.8483337Z #21 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
2021-03-01T13:06:07.6660224Z #21 pushing layers 3.3s done
2021-03-01T13:06:07.6663629Z #21 pushing manifest for ***.dkr.ecr.***.amazonaws.com/bug-report:buildkit-dfa6ddf4705485840c392c397106932c03bf4961
2021-03-01T13:06:08.4085934Z #21 pushing manifest for ***.dkr.ecr.***.amazonaws.com/bug-report:buildkit-dfa6ddf4705485840c392c397106932c03bf4961 0.7s done
2021-03-01T13:06:08.4088745Z #21 ERROR: failed commit on ref "manifest-sha256:672d30d763ba54a9c7f5169573b554f16efd3e8a25deb60413d259a29b204d76": unexpected status: 400 Bad Request
2021-03-01T13:06:08.4090060Z ------
2021-03-01T13:06:08.4090863Z  > exporting to image:
2021-03-01T13:06:08.4091431Z ------
2021-03-01T13:06:08.4093157Z error: failed to solve: rpc error: code = Unknown desc = failed commit on ref "manifest-sha256:672d30d763ba54a9c7f5169573b554f16efd3e8a25deb60413d259a29b204d76": unexpected status: 400 Bad Request
2021-03-01T13:06:08.4160997Z ##[error]buildx call failed with: error: failed to solve: rpc error: code = Unknown desc = failed commit on ref "manifest-sha256:672d30d763ba54a9c7f5169573b554f16efd3e8a25deb60413d259a29b204d76": unexpected status: 400 Bad Request
```

Funnily enough, if I go to the ECR repository and list the images, **there is** something pushed for both:
```
containerd-dfa6ddf4705485840c392c397106932c03bf4961   01 Mar 2021 15:06:51   380.83MB   sha256:d14dffbdc33c894279a22d6ff71671ab6d243a3b488563ad3360bdca781914a9
<untagged>                                            01 Mar 2021 15:06:50   370.38MB   sha256:2fd9e0a11c41f27986d6c0bc6b949a76b85bb9ade27685c4200b87f1c84e991b
<untagged>                                            01 Mar 2021 15:06:50   380.83MB   sha256:43c4984a930ced6b904820bcce54a85a31235cbd915ba3eaf45c31cc7810cf2b
buildkit-dfa6ddf4705485840c392c397106932c03bf4961     01 Mar 2021 15:06:08   380.83MB   sha256:2eb8f205617af17c0af81e1871b43f2d69a59b3c723a9c5bf61258b1712b9411
```

<img width="2466" alt="Screenshot of the ECR console shxtwing the above text" src="https://user-images.githubusercontent.com/5440007/109502489-bc281980-7aa1-11eb-94d3-6b7c3c71265e.png">

--- a/util/push/push.go
+++ b/util/push/push.go
@@ -43,7 +43,12 @@ func Push(ctx context.Context, sm *session.Manager, sid string, provider content
 	if byDigest {
 		ref = parsed.Name()
 	} else {
-		ref = reference.TagNameOnly(parsed).String()
+		// add digest to ref, this is what containderd uses to choose root manifest from all manifests
+		r, err := reference.WithDigest(reference.TagNameOnly(parsed), dgst)
+		if err != nil {
+			return errors.Wrapf(err, "failed to combine ref %s with digest %s", ref, dgst)
+		}
+		ref = r.String()
 	}
 
 	scope := "push"

